### ISSUE TITLE
CI failure: "RuntimeError: tk.h version (8.6) doesn't match libtk.a version (8.5)"

### ISSUE THREAD

**Issue by kousu** [ivadomed/ivadomed#1132]
### Current behavior

CI is failing on {macOS} x {Python 3.7, Python 3.8}. Python3.6 and 3.9 are unaffected.

### Steps to reproduce

Run [CI](.github/workflows/run_tests.yml) on almost any branch currently and it will fail. Example failures:

* `master`: https://github.com/ivadomed/ivadomed/actions/runs/2340863632
* https://github.com/ivadomed/ivadomed/pull/1125: https://github.com/ivadomed/ivadomed/actions/runs/2295180486
  * Interestingly, this one only failed on 3.7, too
* https://github.com/ivadomed/ivadomed/pull/1126: https://github.com/ivadomed/ivadomed/actions/runs/2276043597
  * Interestingly, this one only failed on 3.7
* https://github.com/ivadomed/ivadomed/pull/1127: https://github.com/ivadomed/ivadomed/actions/runs/2321234595
* https://github.com/ivadomed/ivadomed/pull/1128: https://github.com/ivadomed/ivadomed/actions/runs/2327341106
* https://github.com/ivadomed/ivadomed/pull/1129: https://github.com/ivadomed/ivadomed/actions/runs/2340867212
* https://github.com/ivadomed/ivadomed/pull/1131: https://github.com/ivadomed/ivadomed/actions/runs/2341453544
 
Typical pattern of failures:

![Screenshot 2022-05-17 at 18-15-15 Update publish_package yml · ivadomed_ivadomed@c4c425b](https://user-images.githubusercontent.com/987487/168920668-24545ff3-9c20-4dca-b0ff-f4ac1a020704.png)

## Linked Issues

* Being tracked by GitHub staff at https://github.com/actions/setup-python/issues/402

**Comment by kousu** [comment]
The failures are all throwing the same [exception](https://github.com/ivadomed/ivadomed/runs/6478852629?check_suite_focus=true#step:7:382):

```
self = <tkinter.Tk object .>

    def _loadtk(self):
        self._tkloaded = 1
        global _default_root
        # Version sanity checks
        tk_version = self.tk.getvar('tk_version')
        if tk_version != _tkinter.TK_VERSION:
            raise RuntimeError("tk.h version (%s) doesn't match libtk.a version (%s)"
>                              % (_tkinter.TK_VERSION, tk_version))
E           RuntimeError: tk.h version (8.6) doesn't match libtk.a version (8.5)
```

I looked around for help and found an eclectic mixture of reports of this exact error, some on macOS, others in pyenv (which we don't use):

* https://github.com/jiansoung/issues-list/issues/15
* https://github.com/conda-forge/pillow-feedstock/pull/116
* https://github.com/pyenv/pyenv/issues/1375
* https://github.com/pyenv/pyenv/issues/2109
* https://github.com/pyenv/pyenv/issues/1545
* https://github.com/pyenv/pyenv/issues/1125
* https://github.com/spack/spack/issues/23780
* https://github.com/pyenv/pyenv/issues/94
* https://github.com/NixOS/nixpkgs/issues/10667
* https://github.com/Homebrew/legacy-homebrew/pull/41728

as well as this reported fix in `brew`:

* https://github.com/Homebrew/legacy-homebrew/issues/18657
* https://github.com/Homebrew/legacy-homebrew/pull/41728
* https://github.com/Homebrew/legacy-homebrew/issues/41132
* https://github.com/homebrew/homebrew-core/commit/4262fea34a709f430b0c685b15ab2773cbadf520

>  Fix RuntimeError: tk.h version (8.6) doesn't match libtk.a version (8.5)
>
> The problem was that CPPFLAGS and LDFLAGS for brewed Tk were specified to environment variable but Python (seems) prefer the CPPFLAGS in arguments specified to './configure', mean "-I#{MacOS.sdk_path}/usr/include" version is used.


The confusing thing is that that fix went in in 2015, so why is it showing up again now? Also are the Github runners even using brew's python? Do they have their own python build that accidentally recreated the bug? And why only 3.7 and 3.8?

**Comment by kousu** [comment]
Chonologically, the first failure was 13 days ago:

* https://github.com/ivadomed/ivadomed/pull/1126: https://github.com/ivadomed/ivadomed/actions/runs/2276043597

Then 
* https://github.com/ivadomed/ivadomed/pull/1125 : https://github.com/ivadomed/ivadomed/actions/runs/2276419269 - 12 days ago - 3.7 only
* https://github.com/ivadomed/ivadomed/pull/1125: https://github.com/ivadomed/ivadomed/actions/runs/2276436577  - 12 days ago - 3.7 only
* https://github.com/ivadomed/ivadomed/pull/1125: https://github.com/ivadomed/ivadomed/actions/runs/2295180486 - 8 days ago - 3.7 only
* https://github.com/ivadomed/ivadomed/pull/1127: https://github.com/ivadomed/ivadomed/actions/runs/2321234595 - 4 days ago - 3.7 & 3.8
* https://github.com/ivadomed/ivadomed/pull/1128: https://github.com/ivadomed/ivadomed/actions/runs/2327341106 - 2 days ago - 3.7 & 3.8
* https://github.com/ivadomed/ivadomed/pull/990: https://github.com/ivadomed/ivadomed/actions/runs/2333702359 - 1 day ago, 3.7 & 3.8
* https://github.com/ivadomed/ivadomed/pull/1129: https://github.com/ivadomed/ivadomed/actions/runs/2336549525 - today - 3.7 & 3.8
* `master`: https://github.com/ivadomed/ivadomed/actions/runs/2340863632 - today - 3.7 & 3.8
* https://github.com/ivadomed/ivadomed/pull/1129: https://github.com/ivadomed/ivadomed/actions/runs/2340867212 - today - 3.7 & 3.8
* https://github.com/ivadomed/ivadomed/pull/1131: https://github.com/ivadomed/ivadomed/actions/runs/2341453544 - today - 3.7 & 3.8

This smells like a bug in GitHub's hands, especially since it progressed from consistently only being 3.7 to consistently being 3.7 and 3.8. I think they've deployed broken python environments somehow.

**Comment by kousu** [comment]
Interesting, I didn't know this but though [tkinter is a part of cpython](https://docs.python.org/3/library/tkinter.html), but it's packaged separately by Ubuntu:

```
$ apt-cache search python3-tk
python3-tk - Tkinter - Writing Tk applications with Python 3.x
```

Maybe macOS is in a similar situation? And that's prone to breakage like this?

**Comment by kousu** [comment]
A tip from some of those threads: if you do have Tk installed, then this will load and test it:

```
python -m tkinter -c 'tkinter._test()'
```

![Screenshot_20220517_185557](https://user-images.githubusercontent.com/987487/168925452-cff9eedb-63b6-493a-a845-8355e08cb100.png)

If it is GitHub's fault, we should be able to see that by making a fresh CI script that has nothing to do with ivadomed...

**Comment by kousu** [comment]
I think on our end we should figure out why matplotlib is defaulting to Tk. It _should_ be in a headless environment, so it shouldn't be trying to do any GUI things at all. At most, it should be using [`Agg`](https://matplotlib.org/stable/users/explain/backends.html) and rendering to .png.

I think I understand now the failure pattern: [matplotlib explains](https://matplotlib.org/stable/users/explain/backends.html)

>  The first usable backend in the following list is selected: MacOSX, QtAgg, GTK4Agg, Gtk3Agg, TkAgg, WxAgg, Agg. The last, Agg, is a non-interactive backend that can only write to files. It is used on Linux, if Matplotlib cannot connect to either an X display or a Wayland display.

which means it's doing something like

```
for backend in ['MacOS', 'QtAgg', ..., 'TkAgg', ..., 'Agg']:
  try:
    import(backend)
    break
  except InitError:
    pass
else:
  raise 'No Usable Backend'
```

i.e. it's trying each backend in turn until one works. But for some reason, trying to init TkAgg is giving it an error it's not setup to catch -- this low-level library mismatch problem.

If so, we should be able to sidestep the problem by going direct to Agg: https://github.com/ivadomed/ivadomed/pull/1133

**Comment by kousu** [comment]
Here's a different fix than #1133: just downgrade python 3.7 itself (from '3.7' to '3.7.9'):

https://github.com/carlosperate/ubittool/commit/85836202159efb298807a80b4019542df5543d68

If #1133 doesn't work we can try that.
